pkg/kademlia/workers.go

```diff
@@ -0,0 +1,139 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package kademlia
+
+import (
+	"bytes"
+	"context"
+	"math/rand"
+	"sync"
+	"time"
+
+	"storj.io/storj/pkg/node"
+	proto "storj.io/storj/protos/overlay"
+)
+
+const (
+	uncontacted = iota
+	inProgress
+	completed
+)
+
+type chore struct {
+	status int
+	node   *proto.Node
+}
+
+type worker struct {
+	// node id to chore
+	contacted   map[string]*chore
+	mu          *sync.Mutex
+	maxResponse time.Duration
+	cancel      context.CancelFunc
+	nodeClient  node.Client
+	find        proto.Node
+	k           int
+}
+
+func (w *worker) work(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+			w.mu.Lock()
+			var wk *chore
+			for i, v := range w.contacted {
+				if v.status == uncontacted {
+					wk = v
+					w.contacted[i].status = inProgress
+					w.mu.Unlock()
+					break
+				}
+			}
+
+			if wk == nil {
+				w.mu.Unlock()
+				time.AfterFunc(2*w.maxResponse, w.cancel)
+				return nil
+			}
```

Comment:
the mutex semantics above are trippy (but correct afaict). perhaps it would be easier to follow if there was a helper function, even if defined inline, that got the next chore?